src/Build/BackEnd/Components/RequestBuilder/Lookup.cs

```diff
@@ -668,6 +669,7 @@ internal void AddNewItemsOfItemType(string itemType, ICollection<ProjectItemInst
                     itemsToAdd = itemsToAdd.Where(item => !existingItems.Contains(item, ProjectItemInstance.EqualityComparer));
                 }
             }
+            logFunction?.Invoke(itemsToAdd.ToList());
```

Comment:
It feels like we'd be allocating unnecesarily in majority of the cases. Can we just reuse the incomming structure if it wasn't changed?

```suggestion
            logFunction?.Invoke(doNotAddDuplicates ? itemsToAdd.ToList() : itemsToAdd);
```

Reply:
At first, I've done this:
``` logFunction?.Invoke(itemsToAdd.Count() == group.Count ? (List<ProjectItemInstance>)group : itemsToAdd.ToList()); ```
which does the job as far as I can tell. There is a question if it isn't kind of dangerous though:
the function signature allows for any ICollection to be passed. Currently, we're only passing a list (or a list-like collection) that shouldn't have an issue with this. However it seems somewhat iffy to assume that this will always be the case.

Reply:
also please keep in mind that this is the most allocation heavy part of MSBuild so every allocated byte needs to be carefully reasoned about. Items take up the most memory and are a noticeable part of why MSBuild is slow.

Reply:
stylistically I strongly prefer an if statement here instead of a null coalescing operator

Reply:
> this is the most allocation heavy part of MSBuild so every allocated byte needs to be carefully reasoned about

I have slightly less concern since this is at execution time rather than eval, but of course that's no reason not to think about and improve allocations.